src/OVAL/oval_session.c

```diff
@@ -0,0 +1,463 @@
+		oscap_free(path_clone);
+		return 1;
+	}
+	oscap_free(path_clone);
+
+	oval_agent_set_product_name(session->sess, (char *)oscap_productname);
+	return 0;
+}
+
+int oval_session_evaluate_id(struct oval_session *session, char *probe_root, const char *id, oval_result_t *result)
+{
+	__attribute__nonnull__(session);
+
+	if (probe_root) {
+		if (setenv("OSCAP_PROBE_ROOT", probe_root, 1) != 0) {
+			oscap_seterr(OSCAP_EFAMILY_OSCAP, "Failed to set the OSCAP_PROBE_ROOT environment variable.");
+			return 1;
+		}
+	}
+
+	if (id == NULL) {
+		oscap_seterr(OSCAP_EFAMILY_OVAL, "No OVAL Definion id set.");
+		return 1;
+	}
+
+	if (oval_session_setup_agent(session) != 0) {
+		return 1;
+	}
+
+	oval_agent_eval_definition(session->sess, id);
+	*result = OVAL_RESULT_NOT_EVALUATED;
+	oval_agent_get_definition_result(session->sess, id, result);
+	if (oscap_err()) {
+		return 1;
+	}
+
+	session->res_model = oval_agent_get_results_model(session->sess);
+
+	return 0;
+}
+
+int oval_session_evaluate(struct oval_session *session, char *probe_root, agent_reporter fn, void *arg)
+{
+	__attribute__nonnull__(session);
+
+	if (probe_root) {
+		if (setenv("OSCAP_PROBE_ROOT", probe_root, 1) != 0) {
+			oscap_seterr(OSCAP_EFAMILY_OSCAP, "Failed to set the OSCAP_PROBE_ROOT environment variable.");
+			return 1;
+		}
+	}
+
+	if (oval_session_setup_agent(session) != 0) {
+		return 1;
+	}
+
+	oval_agent_eval_system(session->sess, fn, arg);
+	if (oscap_err()) {
+		return 1;
+	}
+
+	session->res_model = oval_agent_get_results_model(session->sess);
+
+	return 0;
+}
+
+int oval_session_export(struct oval_session *session)
+{
+	__attribute__nonnull__(session);
+
+	struct oval_directives_model *dir_model = NULL;
+	struct oscap_source *result = NULL;		/* OVAL Results */
+	const char *filename = NULL;
+	int ret = 0;
+
+	/* Import OVAL Directives if any */
+	if (session->oval.directives && session->res_model) {
+		/* OVAL Directives can be used only if there are OVAL Resutls and these are
+		 * only available if there is a Results model which mean that either
+		 * evaluation or analyse was performed. */
+		if (dir_model)
+			oval_directives_model_free(dir_model);
+		oval_directives_model_import_source(dir_model, session->oval.directives);
+	}
+
+	/* Get OVAL Results if evaluation or analyse has been done and apply
+	 * directives to them */
+	if (session->res_model && (session->export.results || session->export.report)) {
+		result = oval_results_model_export_source(session->res_model, dir_model, NULL);
+		filename = session->export.results;
+	}
+
+	/* Validate OVAL Results. The 'result' in condition will make sure that there is
+	 * something to validate */
+	if (result && session->validation && session->full_validation) {
+		if (!oval_session_validate(session, result, OSCAP_DOCUMENT_OVAL_RESULTS)) {
+		/* FIXME: if invalid output than add suffix ".invalid" to file name but
+		 * still export and set some error about it ?*/
```

Comment:
FIXMEs are like God. They are everlasting.

We could at least return 1 at the bottom when this happens.